### PR TITLE
Vendor update mimir-prometheus at af2a1cb10c89de496cd4309ac532624b34112d74

### PR DESCRIPTION
**Pull request by krajorama**
#### What this PR does

This is upstream at
https://github.com/prometheus/prometheus/tree/af2a1cb10c89de496cd4309ac532624b34112d74

Changes listed in
https://github.com/grafana/mimir-prometheus/pull/787

Fixed query sharding test `TestQuerySharding_FunctionCorrectness` failing due to https://github.com/prometheus/prometheus/pull/15479.

#### Which issue(s) this PR fixes or relates to

N/A

#### Checklist

- [ ] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/streamingpromql/operators/functions/range_vectors.go
Comment (charleskorn) [inline pkg/streamingpromql/operators/functions/range_vectors.go:1]: What if we introduced something like `InstantVectorSeriesDataIterator`, but for iterating over a `RangeVectorStepData`? I think that would simplify a lot of the logic here, and more clearly separate the iteration logic from the `resets` / `changes` logic.

Comment (jhesketh) [reply pkg/streamingpromql/operators/functions/range_vectors.go:1]: I had thought about that too. I'm happy to do it, but shall we do it as a follow up PR?

**Comment by jesusvazquez** [comment]
I'm trying to add two reverts here https://github.com/grafana/mimir/pull/10203 maybe we could merge the effort @krajorama

**Review by charleskorn — APPROVED** [review]
Approving MQE changes to unblock this, but I'd like to see the change suggested [here](https://github.com/grafana/mimir/pull/10188#discussion_r1877247518) applied in a follow up PR, and I'd also like to see us use the exported functions from https://github.com/prometheus/prometheus/pull/15190 in a follow up PR as well.

File: CHANGELOG.md [hunk 1]
```diff
@@ -103,6 +103,15 @@
 * [BUGFIX] Querier: Fix stddev+stdvar aggregations to treat Infinity consistently. #9844
 * [BUGFIX] Ingester: Chunks could have one unnecessary zero byte at the end. #9844
 * [BUGFIX] OTLP receiver: Preserve colons and combine multiple consecutive underscores into one when generating metric names in suffix adding mode (`-distributor.otel-metric-suffixes-enabled`). #10075
+* [BUGFIX] PromQL: Ignore native histograms in `clamp`, `clamp_max` and `clamp_min` functions. #10136
```
Comment (krajorama) [inline CHANGELOG.md:106]: Note to reviewers: native  histograms is an experimental feature, so these don't count as CHANGE.

File: pkg/frontend/querymiddleware/querysharding_test.go [hunk 2]
```diff
@@ -988,15 +988,8 @@ func TestQuerySharding_FunctionCorrectness(t *testing.T) {
 
 	testsForBoth := []queryShardingFunctionCorrectnessTest{
 		{fn: "count_over_time", rangeQuery: true},
-		{fn: "days_in_month"},
```
Comment (krajorama) [inline pkg/frontend/querymiddleware/querysharding_test.go:991]: Note to reviewers: due to https://github.com/prometheus/prometheus/pull/15479

**Comment by krajorama** [comment]
> I'm trying to add two reverts here #10203 maybe we could combine the effort @krajorama

We'll follow up in separate PR. Let this be pure Prometheus upstream update.

**Review by aknuds1 — APPROVED** [review]

File: CHANGELOG.md [hunk 3]
```diff
@@ -103,6 +103,15 @@
...
+* [BUGFIX] PromQL: Ignore native histograms in `max`, `min`, `stdvar`, `stddev` functions and instead return an info annotation. #10136
```
Comment (jhesketh) [inline CHANGELOG.md:107]: Are these github numbers correct. Shouldn't it be this issue (#10188)?

File: CHANGELOG.md [hunk 4]
```diff
@@ -103,6 +103,15 @@
...
+* [BUGFIX] PromQL: Ignore native histograms when compared to float values with `==`, `!=`, `<`, `>`, `<=`, `>=` and instead return an info annotation. #10136
+* [BUGFIX] PromQL: Return an info annotation if the `quantile` function is used on a float series that does not have `le` label. #10136
```
Comment (jhesketh) [inline CHANGELOG.md:109]: I'm confused, I don't see where this change is sorry?

File: CHANGELOG.md (hunk 3 above)
Comment (krajorama) [reply CHANGELOG.md:107]: Yes, they are correct, https://github.com/grafana/mimir/pull/10136 includes https://github.com/prometheus/prometheus/pull/15245 .

File: CHANGELOG.md (hunk 4 above)
Comment (krajorama) [reply CHANGELOG.md:109]: See https://github.com/grafana/mimir/pull/10136 which  includes https://github.com/prometheus/prometheus/pull/15245 .
That PR 15245 contained a bunch of stuff, I just split it up for readability.

Comment (jhesketh) [reply CHANGELOG.md:109]: But that PR does not contain any changes to the quantile function?

File: CHANGELOG.md (hunk 3 above)
Comment (jhesketh) [reply CHANGELOG.md:107]: Okay thanks. Were these just missed from #10136's changelog?

File: CHANGELOG.md (hunk 4 above)
Comment (krajorama) [reply CHANGELOG.md:109]: promql/engine.go:+3079 ?  This is exactly why I asked @beorn7 to review the changelog, maybe I misunderstood something?

File: CHANGELOG.md (hunk 3 above)
Comment (krajorama) [reply CHANGELOG.md:107]: Yes, I prioritized getting the code in over going through the changelog at the time - we had plenty to catch up on

File: CHANGELOG.md (hunk 4 above)
Comment (jhesketh) [reply CHANGELOG.md:109]: Ah, I see it now, yes, thank you.

**Review by jhesketh — APPROVED** [review]
LGTM, thanks!

**Review by jesusvazquez — APPROVED** [review]
LGTM

File: CHANGELOG.md (hunk 3 above)
Comment (beorn7) [inline CHANGELOG.md:107]: Nit: These are aggregation operators, not functions.

File: CHANGELOG.md [hunk 5]
```diff
@@ -103,6 +103,15 @@
...
+* [BUGFIX] PromQL: Fix `count_values` not taking into account native histograms. #10168
```
Comment (beorn7) [inline CHANGELOG.md:110]: `count_values` _does_ take native histograms into account with the recent fix. See https://github.com/prometheus/prometheus/pull/15422

File: CHANGELOG.md [hunk 6]
```diff
@@ -103,6 +103,15 @@
...
+* [BUGFIX] PromQL: Ignore native histograms in time functions `day_of_month`, `day_of_week`, `day_of_year`, `days_in_month`, `hour`, `minute`, `month` and `year`, which means they no longer any value when encountering a native histograms series. #10188
```
Comment (beorn7) [inline CHANGELOG.md:111]: A word is missing here, like "yield" or "return".

**Review by beorn7 — CHANGES_REQUESTED** [review]
Some small things…

File: CHANGELOG.md (hunk 5 above)
Comment (krajorama) [reply CHANGELOG.md:110]: lol, unfortunate wording :)

File: CHANGELOG.md (hunk 6 above)
Comment (krajorama) [reply CHANGELOG.md]: @beorn7 updated the descriptions, thank you.

**Review by beorn7 — APPROVED** [review]
For NH/PromQL part.

**Pull request merged**